examples/bots/discord_bot.py

```diff
@@ -0,0 +1,201 @@
+        self.vector_storage_local_path = vector_storage_local_path
+        self.top_k = top_k
+        self.return_detailed_info = return_detailed_info
+
+        if not self.token:
+            raise ValueError(
+                "`DISCORD_TOKEN` not found in environment variables. Get it"
+                " here: `https://discord.com/developers/applications`."
+            )
+
+        try:
+            import discord
+        except ImportError:
+            raise ImportError(
+                "Please install `discord` first. You can install it by running"
+                " `python3 -m pip install -U discord.py`."
+            )
+        intents = discord.Intents.default()
+        intents.message_content = True
+        self.client = discord.Client(intents=intents)
+
+        # Register event handlers
+        self.client.event(self.on_ready)
+        self.client.event(self.on_message)
+
+    def run(self) -> None:
+        r"""Start the Discord bot using its token.
+
+        This method starts the Discord bot by running the client with the
+        provided token.
+        """
+        self.client.run(self.token)  # type: ignore[arg-type]
+
+    async def on_ready(self) -> None:
+        r"""This method is called when the bot has successfully connected to
+        the Discord server.
+
+        It prints a message indicating that the bot has logged in and displays
+        the username of the bot.
+        """
+        print(f'We have logged in as {self.client.user}')
+
+    async def on_message(self, message: 'Message') -> None:
+        r"""Event handler for when a message is received.
+
+        Args:
+            message (discord.Message): The message object received.
+        """
+
+        # If the message author is the bot itself,
+        # do not respond to this message
+        if message.author == self.client.user:
+            return
+
+        # If allowed channel IDs are provided,
+        # only respond to messages in those channels
+        if self.channel_ids and message.channel.id not in self.channel_ids:
+            return
+
+        # Only respond to messages that mention the bot
+        if not self.client.user or not self.client.user.mentioned_in(message):
+            return
+
+        user_raw_msg = message.content
+
+        if self.auto_retriever:
+            retrieved_content = self.auto_retriever.run_vector_retriever(
+                query=user_raw_msg,
+                vector_storage_local_path=self.content_input_paths,
+                top_k=self.top_k,
+                return_detailed_info=self.return_detailed_info,
+            )
+            user_raw_msg = '''
+                Here is the query to you: {user_raw_msg}
+                Based on the retrieved content:{retrieved_content}, 
+                answer the query from {name}'''.format(
+                user_raw_msg=user_raw_msg,
+                retrieved_content=retrieved_content,
+                name=message.author.name,
+            )
```

Comment:
The prompt structure and content can also be further polished, you can use chatgpt to do it if you are interested in, but I think it's also ok for now since it's one example use case~
```suggestion
            user_raw_msg = (
                f"Here is the query to you: {user_raw_msg}\n"
                f"Based on the retrieved content: {retrieved_content}, \n"
                f"answer the query from {message.author.name}"
            )
```